library/src/main/java/org/apache/apex/malhar/lib/state/managed/IncrementalCheckpointManager.java

```diff
@@ -133,8 +134,10 @@ protected void transferWindowFiles()
           Map<Long, Map<Slice, Bucket.BucketedValue>> buckets = savedWindows.remove(windowId);
 
           for (Map.Entry<Long, Map<Slice, Bucket.BucketedValue>> singleBucket : buckets.entrySet()) {
-            managedStateContext.getBucketsFileSystem().writeBucketData(windowId, singleBucket.getKey(),
-                singleBucket.getValue());
+            long bucketId = singleBucket.getKey();
+            if (bucketId > latestPurgedTimeBucket) {
```

Comment:
What @tweise mentioned is right that the state will be purged via deleteTimeBucketsLessThanEqualTo. So by having this condition are we trying to avoid data which is not transferred yet for a bucket but is already ready for purging? In that case it makes sense. However adding a test to validate that will be good.
Also, `latestPurgedTimeBucket` seems to be a wrong name for this variable. In here it is the main bucket so it should be`lasterPurgedBucketId`. Also I think it should be transient because I mentioned in my earlier comment that it is being modified by another thread and checkpointed by another.